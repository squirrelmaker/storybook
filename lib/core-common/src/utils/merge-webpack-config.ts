import { Configuration } from 'webpack';

function plugins(
  { plugins: defaultPlugins = [] }: Configuration,
  { plugins: customPlugins = [] }: Configuration
): Configuration['plugins'] {
  return [...defaultPlugins, ...customPlugins];
}

function rules(
  { rules: defaultRules = [] }: Configuration['module'],
  { rules: customRules = [] }: Configuration['module']
): Configuration['module']['rules'] {
  return [...defaultRules, ...customRules];
}

function extensions(
  { extensions: defaultExtensions = [] }: Configuration['resolve'],
  { extensions: customExtensions = [] }: Configuration['resolve']
): Configuration['resolve']['extensions'] {
  return [...defaultExtensions, ...customExtensions];
}

function alias(
  { alias: defaultAlias = {} }: Configuration['resolve'],
  { alias: customAlias = {} }: Configuration['resolve']
): Configuration['resolve']['alias'] {
  return {
    ...defaultAlias,
    ...customAlias,
  };
}

function module(
  { module: defaultModule = { rules: [] } }: Configuration,
  { module: customModule = { rules: [] } }: Configuration
): Configuration['module'] {
  return {
    ...defaultModule,
    ...customModule,
    rules: rules(defaultModule, customModule),
  };
}

function resolve(
  { resolve: defaultResolve = {} }: Configuration,
  { resolve: customResolve = {} }: Configuration
): Configuration['resolve'] {
  return {
    ...defaultResolve,
    ...customResolve,
    alias: alias(defaultResolve, customResolve),
    extensions: extensions(defaultResolve, customResolve),
  };
}

function optimization(
  { optimization: defaultOptimization = {} }: Configuration,
  { optimization: customOptimization = {} }: Configuration
): Configuration['optimization'] {
  return {
    ...defaultOptimization,
    ...customOptimization,
  };
}

export function mergeConfigs(config: Configuration, customConfig: Configuration): Configuration {
  return {
    // We'll always load our configurations after the custom config.
    // So, we'll always load the stuff we need.
    ...customConfig,
    ...config,
    devtool: customConfig.devtool || config.devtool,
    plugins: plugins(config, customConfig),
    module: module(config, customConfig),
    resolve: resolve(config, customConfig),
    optimization: optimization(config, customConfig),
  };
}
