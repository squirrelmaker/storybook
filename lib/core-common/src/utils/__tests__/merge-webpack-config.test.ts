import { Configuration } from 'webpack';
import { mergeConfigs } from '../merge-webpack-config';

const config: Configuration = {
  devtool: 'source-map',
  entry: {
    bundle: 'index.js',
  },
  output: {
    filename: '[name].js',
  },
  module: {
    rules: [{ use: 'r1' }, { use: 'r2' }],
  },
  // For snapshot readability purposes `plugins` attribute doesn't match the correct type
  // @ts-expect-errors
  plugins: ['p1', 'p2'],
  resolve: {
    enforceExtension: true,
    extensions: ['.js', '.json'],
    alias: {
      A1: 'src/B1',
      A2: 'src/B2',
    },
  },
  optimization: {
    splitChunks: {
      chunks: 'all',
    },
    runtimeChunk: true,
  },
};

describe('mergeConfigs', () => {
  it('merges two full configs in one', () => {
    const customConfig: Configuration = {
      profile: true,
      entry: {
        bundle: 'should_not_be_merged.js',
      },
      output: {
        filename: 'should_not_be_merged.js',
      },
      module: {
        noParse: /jquery|lodash/,
        rules: [{ use: 'r3' }, { use: 'r4' }],
      },
      // For snapshot readability purposes `plugins` attribute doesn't match the correct type
      // @ts-expect-errors
      plugins: ['p3', 'p4'],
      resolve: {
        enforceExtension: false,
        extensions: ['.ts', '.tsx'],
        alias: {
          A3: 'src/B3',
          A4: 'src/B4',
        },
      },
      optimization: {
        // For snapshot readability purposes `minimizer` attribute doesn't match the correct type
        // @ts-expect-errors
        minimizer: ['banana'],
      },
    };

    const result = mergeConfigs(config, customConfig);

    expect(result).toMatchSnapshot();
  });

  it('merges partial custom config', () => {
    const customConfig: Configuration = {
      // For snapshot readability purposes `plugins` attribute doesn't match the correct type
      // @ts-expect-errors
      plugins: ['p3'],
      resolve: {
        extensions: ['.ts', '.tsx'],
      },
    };

    const result = mergeConfigs(config, customConfig);

    expect(result).toMatchSnapshot();
  });

  it('merges successfully if custom config is empty', () => {
    const customConfig = {};

    const result = mergeConfigs(config, customConfig);

    expect(result).toMatchSnapshot();
  });
});
