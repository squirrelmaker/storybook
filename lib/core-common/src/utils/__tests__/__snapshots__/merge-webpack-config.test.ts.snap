// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeConfigs merges partial custom config 1`] = `
Object {
  "devtool": "source-map",
  "entry": Object {
    "bundle": "index.js",
  },
  "module": Object {
    "rules": Array [
      Object {
        "use": "r1",
      },
      Object {
        "use": "r2",
      },
    ],
  },
  "optimization": Object {
    "runtimeChunk": true,
    "splitChunks": Object {
      "chunks": "all",
    },
  },
  "output": Object {
    "filename": "[name].js",
  },
  "plugins": Array [
    "p1",
    "p2",
    "p3",
  ],
  "resolve": Object {
    "alias": Object {
      "A1": "src/B1",
      "A2": "src/B2",
    },
    "enforceExtension": true,
    "extensions": Array [
      ".js",
      ".json",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`mergeConfigs merges successfully if custom config is empty 1`] = `
Object {
  "devtool": "source-map",
  "entry": Object {
    "bundle": "index.js",
  },
  "module": Object {
    "rules": Array [
      Object {
        "use": "r1",
      },
      Object {
        "use": "r2",
      },
    ],
  },
  "optimization": Object {
    "runtimeChunk": true,
    "splitChunks": Object {
      "chunks": "all",
    },
  },
  "output": Object {
    "filename": "[name].js",
  },
  "plugins": Array [
    "p1",
    "p2",
  ],
  "resolve": Object {
    "alias": Object {
      "A1": "src/B1",
      "A2": "src/B2",
    },
    "enforceExtension": true,
    "extensions": Array [
      ".js",
      ".json",
    ],
  },
}
`;

exports[`mergeConfigs merges two full configs in one 1`] = `
Object {
  "devtool": "source-map",
  "entry": Object {
    "bundle": "index.js",
  },
  "module": Object {
    "noParse": /jquery\\|lodash/,
    "rules": Array [
      Object {
        "use": "r1",
      },
      Object {
        "use": "r2",
      },
      Object {
        "use": "r3",
      },
      Object {
        "use": "r4",
      },
    ],
  },
  "optimization": Object {
    "minimizer": Array [
      "banana",
    ],
    "runtimeChunk": true,
    "splitChunks": Object {
      "chunks": "all",
    },
  },
  "output": Object {
    "filename": "[name].js",
  },
  "plugins": Array [
    "p1",
    "p2",
    "p3",
    "p4",
  ],
  "profile": true,
  "resolve": Object {
    "alias": Object {
      "A1": "src/B1",
      "A2": "src/B2",
      "A3": "src/B3",
      "A4": "src/B4",
    },
    "enforceExtension": false,
    "extensions": Array [
      ".js",
      ".json",
      ".ts",
      ".tsx",
    ],
  },
}
`;
